x-pack/plugins/security_solution/server/lib/telemetry/task.ts

```diff
@@ -0,0 +1,121 @@
+ */
+import moment from 'moment';
+import { Logger } from 'src/core/server';
+import {
+  ConcreteTaskInstance,
+  TaskManagerSetupContract,
+  TaskManagerStartContract,
+} from '../../../../task_manager/server';
+import { TelemetryEventsSender, TelemetryEvent } from './sender';
+
+export const TelemetryDiagTaskConstants = {
+  TIMEOUT: '1m',
+  TYPE: 'security:endpoint-diagnostics',
+  INTERVAL: '5m',
+  VERSION: '1.0.0',
+};
+
+export class TelemetryDiagTask {
+  private readonly logger: Logger;
+  private readonly sender: TelemetryEventsSender;
+
+  constructor(
+    logger: Logger,
+    taskManager: TaskManagerSetupContract,
+    sender: TelemetryEventsSender
+  ) {
+    this.logger = logger;
+    this.sender = sender;
+
+    taskManager.registerTaskDefinitions({
+      [TelemetryDiagTaskConstants.TYPE]: {
+        title: 'Security Solution Telemetry Diagnostics task',
+        timeout: TelemetryDiagTaskConstants.TIMEOUT,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              const executeTo = moment().utc().toISOString();
+              const executeFrom = this.getLastExecutionTimestamp(
+                executeTo,
+                taskInstance.state?.lastExecutionTimestamp
+              );
+              await this.runTask(taskInstance.id, executeFrom, executeTo);
+
+              return {
+                state: {
+                  lastExecutionTimestamp: executeTo,
+                },
+              };
+            },
+            cancel: async () => {},
+          };
+        },
+      },
+    });
+  }
+
+  public getLastExecutionTimestamp(executeTo: string, lastExecutionTimestamp?: string) {
+    if (lastExecutionTimestamp === undefined) {
+      this.logger.debug(`No last execution timestamp defined`);
+      return moment(executeTo).subtract(5, 'minutes').toISOString();
+    }
+
+    if (moment(executeTo).diff(lastExecutionTimestamp, 'minutes') >= 10) {
+      this.logger.debug(`last execution timestamp was greater than 10 minutes`);
+      return moment(executeTo).subtract(10, 'minutes').toISOString();
+    }
+
+    return lastExecutionTimestamp;
+  }
+
+  public start = async (taskManager: TaskManagerStartContract) => {
+    try {
+      await taskManager.ensureScheduled({
+        id: this.getTaskId(),
+        taskType: TelemetryDiagTaskConstants.TYPE,
+        scope: ['securitySolution'],
+        schedule: {
+          interval: TelemetryDiagTaskConstants.INTERVAL,
+        },
+        state: {},
+        params: { version: TelemetryDiagTaskConstants.VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task, received ${e.message}`);
+    }
+  };
+
+  private getTaskId = (): string => {
+    return `${TelemetryDiagTaskConstants.TYPE}:${TelemetryDiagTaskConstants.VERSION}`;
+  };
+
+  public runTask = async (taskId: string, searchFrom: string, searchTo: string) => {
```

Comment:
This isn't an objection in the code, just a question: If there are multiple Kibana instances, it is possible for this task to be running simultaneously in multiple instances? Or does the task manager ensure that only one execution can happen at any given time?

Reply:
@stevewritescode Good question. The taskManager uses a distributed model which ensures that only a single Kibana instance will run the task. Each Kibana instance polls for new tasks on an interval and attempts to "claim" tasks that are ready to fire. If another Kibana instance tries to claim the same task, only one will succeed and the others will get a 409 Conflict as per OCC. If I'm remembering all this correctly. :)